[wasm-mt] Support async JS interop on threadpool threads

lambdageek opened the pull request:
This is Part 4 of https://github.com/dotnet/runtime/issues/84489 - initial runtime support for async JS interop on threadpool threads in multi-threaded WebAssembly.

This depends on [Part1](https://github.com/dotnet/runtime/pull/84490), [Part2](https://github.com/dotnet/runtime/pull/84491), [Part3](https://github.com/dotnet/runtime/pull/84492) and the [Emscripten Bump PR](https://github.com/dotnet/runtime/pull/83998).

This is equivalent to the abandoned #83838.

Conceptually there are a few things here:

1. A mechanism for the native runtime to start some threads as "exitable" so that we don't clean up when they return from their thread start function. (we will clean up when the pthreads TLS dtor runs)
2. A change to make `JSHostImplementation.s_csOwnedObjects` a thread-static in multithreaded builds.  This is because the map keys on a small integer handle that is allocated per-thread on the JS side as a handle for JS objects that are to be kept alive because managed code depends on them. (This is needed in general, but also makes the new smoke test work)
3. A version of the PortableThreadPool.WorkerThread that starts as an exitable thread and uses asynchronous callbacks to wait for available threadpool work items and returns to the JS event loop periodically in order to allow JS promises to settle.
4. A smoke test that does a background JS `fetch` and some async callbacks on it.

lambdageek committed:
[wasm-mt] Full JSInterop on threadpool workers

[wasm][threads] flip YieldFromDispatchLoop; specialize PortableThreadPool.WorkerThread

[mono] Implement a LifoJSSemaphore

This is a LIFO semaphore with an asynchronous wait that triggers
callbacks on the JS event loop in case of Release or timeout.

Make managed LowLevelJSSemaphore

copy-paste PortableThreadPool.WorkerThread for threaded WASM

no changes yet. just copying verbatim to a separate file

fixup native code for lifo semaphore

fixup managed code for LowLevelJSSemaphore

Implement PortableThreadPool loop using semaphore callbacks

manage emscripten event loop from PortableThreadPool.WorkerThread

make sure to keep the thread alive after setting up the semaphore
wait.
Cleanup the thread when exiting

FIXME: thread equality assertion in timeout callback

XXX REVERT ME - minimal async timeout test

BUGFIX: &wait_entry ===> wait_entry

nit: log thread id as hex in .ts

Match the C logging for easier grepping

XXX minimal sample - fetch on a background thread works

fix non-wasm non-threads builds

Add WebWorkerEventLoop internal class to managed event loop keepalive

Don't explicitly call UnwindToJs as it doesn't know about managed
code.
Also avoid mono_thread_exit as that also won't necessarily clean up
after the interpreter

Start threadpool threads with keepalive checks

Add a flag to mono's thread start wrappers to keep track of threads
that may not want cleanup to run after the Start function returns.

Use the flag when starting threadpool threads.

HACK: kind of work around the emscripten_runtime_keepalive_push/pop no-op

Keep a thread local counter in mono. that we "know" will be right for
us and manually call unwind_to_js and thread_exit.

This is super-fragile since we don't know what emscripten internals
might be trying to manipulate the keepalive count and also we are
exiting the thread with active managed frames, so we might be skipping
finally clauses and possibly leaking interpreter memory.

This is mainly meant to keep work going on this branch and not
something we necessarily want to commit

support JS Semaphore with --print-icall-table cross compiler

fixes smoketest on Release builds

make minimal FetchBackground sample more like a unit test

Share PortableThreadPool.WorkerThread common code

Share the code between non-browser implementations and
browser+threads.

The differences are just in how the work loop is started and implemented

make both kinds of lifo semaphore share a base struct

Unify LowLevelLifoSemaphore for normal and async waiting

WebWorkerEventLoop: remove dead code, update comments

remove unused arg from async wait semaphore

don't need both user data and a gchandle

rename native semaphore to LifoSemaphoreAsyncWait

prefix functions with mono_lifo_semaphore_asyncwait_

Rename managed file to LowLevelLifoSemaphore.AsyncWait.Browser.Threads.Mono.cs

Remove unnecessary indirections and allocations from managed AsyncWait semaphore

fix non-browser+threads builds

Keep track of unsettled JS interop promises in threadpool workers

Set WorkerThread.IsIOPending when the current thread has unsettled JS
interop promises.

When IsIOPending is true, the worker will not exit even if it has no
more work to do.  Instead it will repeatedly wait for more work to
arrive or for all promises to settle.

change minimal sample's fetch helper to artificially delay

the delay is longer that the threadpool worker's semaphore timeout, in
order to validate that the worker stays alive while there are
unsettled promises

lambdageek committed:
[wasm-mt] Add background interop to smoketest

lambdageek committed:
update to use the LowLevelLifoAsyncWaitSemaphore

lambdageek committed:
adjust to renamed PortableThreadPool helper methods

lambdageek committed:
adjust to renamed WebWorkerEventLoop.HasJavaScriptInteropDependents

lambdageek committed:
extend and rationalize the smoke test a bit

Add a test that just starts a thread and asserts that it has a
different thread ID than the main thread.  This should allow us to
rule out accidentally having the test pass on a single-threaded runtime

lambdageek committed:
remove old-Emscripten workaround hack

lambdageek commented:
@pavelsavara @kg this is probably ok to look at once Part 2 and Part 3 go in.

@vargaz or @lateralusX: could you take a look at the changes in `threads.c` around `create_thread` and `start_wrapper_internal` for the `MONO_THREAD_CREATE_FLAGS_RETURNS_TO_JS_EVENT_LOOP` stuff

pavelsavara commented on src/mono/wasm/runtime/pthreads/worker/index.ts:
```diff
@@ -114,7 +114,7 @@ function onMonoConfigReceived(config: MonoConfigInternal): void {
     const self = pthread_self;
     mono_assert(self !== null && self.pthread_id == pthread_id, "expected pthread_self to be set already when attaching");
-    console.debug("MONO_WASM: attaching pthread to runtime", pthread_id);
+    console.debug("MONO_WASM: attaching pthread to runtime 0x" + pthread_id.toString(16));
```
Please add `if (runtimeHelpers.diagnosticTracing) ` ...
Ideally all places which are missing it.

lambdageek committed:
hide some debug output

lambdageek commented:
after rebasing on main. I'm getting some thread affinity errors in one of the smoke tests here (from https://github.com/dotnet/runtime/pull/84981).  Investigating...

lambdageek commented:
@pavelsavara I'm going to need to make `JSHostImplementation.s_csOwnedObjects` a `[ThreadStatic]`: the affinity assertion was a false positive because when we do the lookup in we get back some `jsHandle` value that is a small integer value - which is thread-local - the main thread could have already used that same integer for its own object.

https://github.com/dotnet/runtime/blob/9df61d9440f7bba9d4ccebcfe8717de5db63445d/src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Marshaling/JSMarshalerArgument.Task.cs#L297-L299

So the lookup in the cache in `CreateCSOwnedProxy` returns the `JSObject` that was created for the main thread for something else.

(I noticed that `s_csOwnedObjects` is also used from Legacy. I'll make sure the legacy interop works as it did before)
the passed-in

lambdageek commented:
There's also an interesting thing happenign in the smoke test after everything sits idle for a while

```
error: [MONO] * Assertion at /Users/alklig/work/dotnet-runtime/runtime/src/mono/mono/utils/mono-threads-coop.c:99, condition `info' not met, function:check_info, (null) Cannot enter GC unsafe region if the thread is not attached

    at mono_wasm_stringify_as_error_with_stack (dotnet.js:848:22)
    at mono_wasm_trace_logger (dotnet.js:867:31)
    at wasm_trace_logger (dotnet.wasm:0x1173c)
    at eglib_log_adapter (dotnet.wasm:0x56783)
    at monoeg_g_logstr (dotnet.wasm:0x50869)
    at monoeg_g_logv_nofree (dotnet.wasm:0x507b6)
    at monoeg_assertion_message (dotnet.wasm:0x508e9)
    at check_info (dotnet.wasm:0x66569)
    at mono_threads_enter_gc_unsafe_region_unbalanced_with_info (dotnet.wasm:0x669a2)
    at mono_threads_enter_gc_unsafe_region_with_info (dotnet.wasm:0x66913)
mono_wasm_trace_logger	@	dotnet.js:867
$wasm_trace_logger	@	dotnet.wasm:0x1173c
$eglib_log_adapter	@	dotnet.wasm:0x56783
$monoeg_g_logstr	@	dotnet.wasm:0x50869
$monoeg_g_logv_nofree	@	dotnet.wasm:0x507b6
$monoeg_assertion_message	@	dotnet.wasm:0x508e9
$check_info	@	dotnet.wasm:0x66569
$mono_threads_enter_gc_unsafe_region_unbalanced_with_info	@	dotnet.wasm:0x669a2
$mono_threads_enter_gc_unsafe_region_with_info	@	dotnet.wasm:0x66913
$mono_threads_enter_gc_unsafe_region_internal	@	dotnet.wasm:0x668c9
$mono_runtime_invoke	@	dotnet.wasm:0x15b366
$mono_wasm_invoke_method_bound	@	dotnet.wasm:0x119ee
(anonymous)	@	dotnet.js:18887
ccall	@	dotnet.js:24753
(anonymous)	@	dotnet.js:24764
invoke_method_and_handle_exception	@	dotnet.js:10976
runtimeHelpers.javaScriptExports.release_js_owned_object_by_gc_handle	@	dotnet.js:11126
teardown_managed_proxy	@	dotnet.js:7180
_js_owned_object_finalized	@	dotnet.js:7190
```

which means we tried to run some C# code after the pthread was already done.   So we might need to keep the worker thread alive while these js_owned_object destructors are still armed.

I think this just started after I made the `s_csOwnedObjects` change locally, which probably means my smoke test was using the js interop layer in an incorrect way before (and so maybe the wrong js owned objects were getting used on the managed side)

lambdageek committed:
smoke test: dispose of the ImportAsync result after the task is done

lambdageek committed:
[wasm-mt] make JSHostImplementation.s_csOwnedObjects ThreadStatic

The integer jsHandles are not global - they are essentially indices
into a JS array.
So the mapping from a jsHandle to a JSObject must be per-thread.

This fixes the thread affinity assertions in the smoketest (which were
false positives - we looked up a worker's jsHandle and got back the
main thread's JSObject - and then asserted that it was accessed from
the wrong thread)

lambdageek commented:
Re: `_js_owned_object_finalized`.

So there's two issues:
1. I'm not 100% sure which C# object the JS code was still holding on to.  I thought everything in the samples was disposed promptly (certainly the Tasks all call `teardown_managed_proxy`, AFAICT).
2. I think what we might have to do here is to forward all the calls to `release_js_owned_object_by_gc_handle` back to the main thread if the current worker isn't attached anymore.  If there's any thread running on the worker, we can invoke the managed cleanup code - it doesn't actually care what thread it runs on.  But if the user code leaked a JS owned C# object and the thread already detached, we either have to retain the C# object forever, or we can forward the call back to the main thread and have it clean up for us (we can use our own message channel from the worker back to the main thread, or else use some emscripten API - I think either one would work).

A minor complication is that I didn't make a friendly way for the JS code to detect if a pthread detached (you can poke into the emscripten Module to get that info, but I don't have a friendly wrapper).

pavelsavara commented:
> I'm going to need to make `JSHostImplementation.s_csOwnedObjects` a `[ThreadStatic]`

Makes sense. And you could remove `lock (s_csOwnedObjects)` in places.

> I think what we might have to do here is to forward all the calls to `release_js_owned_object_by_gc_handle` back to the main thread

I would prefer if we rather just release all those proxies when we detach the worker. 
- for the proxies of C# objects, we would reset the `GCHandle` to `-1` and if some stray `setTimeout` callback uses it, we will throw.
- for the proxies of C# functions, it would take bit more effort. We essentially mark it disposed and it would throw if you try to call it.
- same for proxies in the other direction.
- at the end of detach, we should assert that have empty `FinalizationRegistry` of that worker. 
- at the end of detach, we should assert that s_csOwnedObjects is empty
The next time the same worker is attached, it should have no old proxies either way.

Do you want me to implement it ?

lambdageek commented:
> I would prefer if we rather just release all those proxies when we detach the worker.
> 
> * for the proxies of C# objects, we would reset the `GCHandle` to `-1` and if some stray `setTimeout` callback uses it, we will throw.
> * for the proxies of C# functions, it would take bit more effort. We essentially mark it disposed and it would throw if you try to call it.
> * same for proxies in the other direction.
> * at the end of detach, we should assert that have empty `FinalizationRegistry` of that worker.
> * at the end of detach, we should assert that s_csOwnedObjects is empty
>   The next time the same worker is attached, it should have no old proxies either way.
> 
> Do you want me to implement it ?

Yea I think this approach makes sense.  Let's hold off on implementing.  I need to think a little bit about the runtime side: 
- we need a JS callback or event when a pthread detaches
- but also the FinalizationRegistry callback invokes managed code, so I need to check when is the last moment when the thread is still attached enough that we can run some managed code

lambdageek commented:
/azp run runtime-wasm

lambdageek committed:
remove locking on JSHostImplementation.CsOwnedObjects

In single-threaded wasm, there is no need to lock since there is only
one caller at a time.

In multi-threaded wasm, the underlying dictionary is thread-static

lambdageek commented:
I removed the locks for `JSHostImplementation.CsOwnedObjects` (and made it a property and renamed it since it's accessed from outside the class)

lambdageek committed:
Merge remote-tracking branch 'origin/main' into pieces-wasm-threadpool-worker-4-async-threadpool-worker

lambdageek committed:
Merge remote-tracking branch 'origin/main' into pieces-wasm-threadpool-worker-4-async-threadpool-worker

lambdageek commented:
@lateralusX @vargaz When you have time, could please you take a look at the thread creation changes in this PR (threads.c)?

lambdageek commented:
From Zoltan:

> I'd suggest reducing the amount of ifdefs.  
> adding flags and fields to StartInfo doesn't need an ifdef imho.

From myself:

> kind of unhappy with how much Emscripten stuff leaked into threads.c.  gonna move some of it out to mono-threads-wasm.c and make some "cross platform" sounding names for it.  "has_extern_eventloop" or something

pavelsavara reviewed (approved):
`System.Runtime.InteropServices.JavaScript` and TypeScript changes look good to me.
I don't love the messy `Wasm.Browser.Threads.Minimal.Sample`, but I'm happy to deal with that in the future.

lambdageek committed:
[threads] make the "external eventloop" platform independent

It only does something on WASM, but in principle if other platforms
allow us to run some code after returning from a thread start
function, we could do it there, too.

kg commented on src/mono/mono/metadata/threads.c:
```diff
@@ -1247,13 +1243,11 @@ start_wrapper_internal (StartInfo *start_info, gsize *stack_ptr)
-	if (returns_to_js_event_loop) {
-		/* if the thread wants to stay alive, don't clean up after it */
-		if (emscripten_runtime_keepalive_check())
+	if (G_UNLIKELY (external_eventloop)) {
```
Is the G_UNLIKELY here important or necessary? Just curious

lambdageek replied:
Neither

lateralusX replied:
It is also a hint to compilers that supports it to put the branch out of line assisting the branch predictors heuristics, so it should then end up with a conditional jmp to an out of line code block when true, but branch predictor will predict it to not be taken (normal default heuristics for forward branches).

kg commented on src/mono/mono/metadata/threads.c:
```diff
@@ -4947,7 +4935,7 @@ ves_icall_System_Threading_Thread_StartInternal (MonoThreadObjectHandle thread_h
 	// HACK: threadpool threads can return to the JS event loop
 	// WISH: support this for other threads, too
```
Maybe this should be TODO or FIXME instead of wish so it's easier to find in codebase audits. I assume we're going to get to it eventually either way though

lambdageek replied:
I should probably just add a managed `internal` flag on `MonoThreadObject` and get rid of this hack entirely.  I've more or less convinced myself that we will eventually need a wasm-specific API to set something like this on threads created with `var t = new Thread(...); t.Start()`, anyway

kg reviewed (approved):
Looks right to me based on my understanding of the codebase

kg commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.cs:
```diff
@@ -15,7 +15,20 @@ internal static partial class JSHostImplementation
+#endif
+        private static Dictionary<int, WeakReference<JSObject>>? s_csOwnedObjects;
+
+        public static Dictionary<int, WeakReference<JSObject>> CsOwnedObjects
```
It might be good at some point to rename this to something like ThreadCsOwnedObjects to make it clear that it is objects owned by the current thread only and not all threads, but that's not important

lambdageek committed:
fix wasi and singlethreaded browser-wasm

lambdageek committed:
Add a Thread.HasExternalEventLoop managed property

Set it from WebWorkerEventLoop.StartExitable.
In native code, use it to set the
`MONO_THREAD_CREATE_FLAGS_EXTERNAL_EVENTLOOP` flag when starting the thread.

lambdageek committed:
rename JSHostImplementation.ThreadCsOwnedObjects

(used to be CsOwnedObjects)

Rename to make it clear that it's objects owned by the current thread,
not the runtime globally

lateralusX commented on src/mono/mono/metadata/threads.c:
```diff
@@ -1266,9 +1275,16 @@ start_wrapper (gpointer data)
 	/* Run the actual main function of the thread */
 	res = start_wrapper_internal (start_info, (gsize*)info->stack_end);
 
+	if (G_UNLIKELY (external_eventloop)) {
```
How will this thread be used after it has returned from its thread function? Will we use it to run other code (JS event loop), "impersonating" the thread, like having events that should run managed code, referencing this thread, so when run picked up in JS even loop, it will still be able to complete those events, even if the original thread execution has stopped?

lambdageek replied:
Yes, exactly right.

In order for JS promises to resolve, the JS eventloop must get a chance to run.  So if we create a worker thread with `Task.Run` and  `await` on an imported JS promise (which gets exposed as a `Task<T>` in C#) the worker thread must be able to run the JS eventloop and then invoke the managed code once the JS promise is finished.

The earlier PRs (see the PR description) refactored the managed threadpool worker code to use callbacks that are registered with the JS event loop in order to signal a threadpool worker when it has work available.  The worker stays alive as long as there are unsettled JS promises (that are awaited by managed code) and then either exits (normal `pthread_exit` shuts down these "external eventloop" pthreads same as normal pthreads) or picks up additional work - depending on what the threadpool algorithm says.

lambdageek replied:
> If so, will the thread move over to safe mode since it will still be attached to runtime, but not able to react to any suspend requests etc?

yes it should be in GC Safe when it's in the JS loop (I'll do a checked build this afternoon to verify.  but I'm pretty sure that actually happens as part of the normal thread state management for the managed thread start function invocation. when it returns we will go to gc safe)

lateralusX replied:
> In order for JS promises to resolve, the JS eventloop must get a chance to run. So if we create a worker thread with `Task.Run` and `await` on an imported JS promise (which gets exposed as a `Task<T>` in C#) the worker thread must be able to run the JS eventloop and then invoke the managed code once the JS promise is finished.

So then we keep a reference count or similar making sure the managed thread gets detached when there are no more outstanding references to the thread, correct? Sounded like we depend on TLS destructors to do that, so does code reacting to callback from JS eventloop, invoke managed code and then check refcount and if that is 0, close native thread, that in turn will trigger TLS destructors, detaching thread from runtime?

lambdageek replied:
Correct

lambdageek replied:
Added a gc safe assertion when we're exiting to the external event loop

lateralusX reviewed (approved):
Changes to native code LGTM!

lambdageek committed:
[checked] assert GC Safe mode, when returning to external eventloop
